File: leetcode/trie/677. Map Sum Pairs/py/main.py
class MapSum:

    def __init__(self):
        self.tree = Trie()
        self.pairs= {}

    def insert(self, key: str, val: int) -> None:
        if key in self.pairs:
          self.tree.insert(key, val - self.pairs[key])
          self.pairs[key] = val
          return
        
        self.pairs[key] = val
        self.tree.insert(key, val)

    def sum(self, prefix: str) -> int:
        return self.tree.startsWith(prefix)

class Trie:

    def __init__(self):
        self.root = Node()        

    def insert(self, word: str, value) -> None:
        current = self.root
        
        for w in word:
            if w not in current.children:
                current.children[w] = Node()
            current = current.children[w]
            current.value += value
        
    def startsWith(self, prefix: str) -> bool:
        current = self.root

        for w in prefix:
            if w not in current.children:
                return 0
            current = current.children[w]
            
        return current.value
        
class Node:
    def __init__(self, ):
      self.children = {}
      self.value = 0

File: leetcode/trie/677. Map Sum Pairs/js/main.js
// Time: каждая операция вставки - это O(K), где K - длина ключа. Каждая сумма операции O(K).
// Space: O(K)
class MapSum {
  constructor() {
    this.root = new TrieNode();
    this.map = new Map();
  }

  insert(key, value) {
    let delta = value - (this.map.get(key) || 0);

    let current = this.root;
    this.map.set(key, value);

    current.sum += delta;

    for (let k of key) {
      if (!(k in current.children)) {
        current.children[k] = new TrieNode();
      }
      current = current.children[k];
      current.sum += delta;
    }

    return null;
  }

  sum(prefix) {
    let current = this.root;

    for (let p of prefix) {
      if (!(p in current.children)) {
        return 0;
      }
      current = current.children[p];
    }

    return current.sum;
  }
}

class TrieNode {
  constructor() {
    this.children = {};
    this.sum = null;
  }
}
